### DISCUSSION TITLE
Is there anyway to add Nonlinear crystal or Metal in WarpX simulation?

### DISCUSSION THREAD

**Discussion by jiecaix** [BLAST-WarpX/warpx#5168] [Q&A] [answered]
I noticed that there exist "macroscopic" medium in warpx, so I wonder is it possible to add some metal such as Au, Cu or Nonlinear crystal such as LN in WarpX simulation？

**Comment by ax3l** [comment]
Hi @zhazhajust,

We currently only have an implementation for isotropic, linear macroscopic media. One could extend this (we don't have a project on it, but contributions welcome).

**Comment by RemiLehe** [comment, marked as answer]
Regarding metals, WarpX has two ways to model them:
- As a perfectly conducting boundary: in that case, you would use the WarpX [Embedded Boundary](https://warpx.readthedocs.io/en/latest/usage/parameters.html#embedded-boundary-conditions) functionality.
- As a medium with finite conductivity: in that case, you would use the conductivity parameter of the [macroscopic medium](https://warpx.readthedocs.io/en/latest/usage/parameters.html#maxwell-solver-macroscopic-media): `macroscopic.sigma_function(x,y,z)`. Note that this conductivity would be frequency-independent.
